src/api/MainSolver.cc

```diff
@@ -559,12 +560,32 @@ void MainSolver::Preprocessor::initialize() {
     substitutions.push();
 }
 
+void MainSolver::Preprocessor::prepareForProcessingFrame(std::size_t frameIndex) {
+    assert(frameIndex < virtualFrameCount);
+    while (actualFrameCount <= frameIndex) {
+        pushInternal();
+    }
+}
 void MainSolver::Preprocessor::push() {
+    ++virtualFrameCount;
+}
+
+void MainSolver::Preprocessor::pop() {
+    --virtualFrameCount;
+    if (virtualFrameCount < actualFrameCount) {
+        assert(virtualFrameCount + 1 == actualFrameCount);
+        popInternal();
+    }
```

Comment:
what about:
```
if (virtualFrameCount >= actualFrameCount) return;
popInternal();
assert(virtualFrameCount == actualFrameCount);
```
I think that the final state is more important than an intermediate.